tls/s2n_ktls.c

```diff
@@ -0,0 +1,173 @@
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     *
+     * - Confirm application is has not using custom I/O
+     * - Confirm kTLS isn't enabled already
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE(!conn->managed_send_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_send_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE(!conn->managed_recv_io, S2N_ERR_KTLS);
+            RESULT_ENSURE(!conn->ktls_recv_enabled, S2N_ERR_KTLS);
+            break;
+        case S2N_KTLS_MODE_DISABLED:
+        case S2N_KTLS_MODE_DUPLEX:
+            RESULT_BAIL(S2N_ERR_T_INTERNAL);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        /* retrieve the send fd */
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+
+    /* register the tls ULP */
+    int fd;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+    RESULT_GUARD_POSIX(setsockopt(fd, SOL_TCP, TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE));
+
+    /* TODO configure keys */
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * kTLS has been enabled on the socket. Errors are likely to be fatal and
+ * unrecoverable.
+ */
+S2N_RESULT s2n_ktls_configure_connection(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* TODO mark kTLS enabled on the connection */
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_enable(struct s2n_connection *conn)
+{
+    if (s2n_result_is_error(s2n_ktls_validate(conn))) {
+        return S2N_RESULT_OK;
+    }
+
+    if (s2n_result_is_ok(s2n_ktls_configure_socket(conn, S2N_KTLS_MODE_RECV))) {
+        RESULT_ENSURE_OK(s2n_ktls_configure_connection(conn, S2N_KTLS_MODE_RECV), S2N_ERR_KTLS);
+    }
+
+    if (s2n_result_is_ok(s2n_ktls_configure_socket(conn, S2N_KTLS_MODE_SEND))) {
+        RESULT_ENSURE_OK(s2n_ktls_configure_connection(conn, S2N_KTLS_MODE_SEND), S2N_ERR_KTLS);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+#else
+
+S2N_RESULT s2n_ktls_enable(struct s2n_connection *conn)
+{
+    /* kTLS feature is not supported on this platform */
+    return S2N_RESULT_OK;
```

Comment:
Should this return `S2N_RESULT_OK` if it's not supported? For example, `s2n_config_set_status_request_type` returns a `S2N_ERR_OCSP_NOT_SUPPORTED` error if you try to request OCSP stapling and it isn't supported. Should there be a similar kTLS not supported error?